rasterio/_io.pyx

```diff
@@ -1063,6 +1077,54 @@ cdef class DatasetReaderBase(DatasetBase):
         # generator implemented in sample.py.
         return sample_gen(self, xy, indexes=indexes, masked=masked)
 
+    def statistics(self, bidx, approx=False, clear_cache=False):
+        """Get min, max, mean, and standard deviation of a raster band.
+
+        Parameters
+        ----------
+        bidx : int
+            The band's index (1-indexed).
+        approx : bool, optional
+            If True, statistics will be calculated from reduced
+            resolution data.
+        clear_cache : bool, optional
+            If True, saved stats will be deleted and statistics will be
+            recomputed. Requires GDAL version >= 3.2.
+
+        Returns
+        -------
+        Statistics
+
+        Notes
+        -----
+        GDAL will preferentially use statistics kept in raster metadata
+        like images tags or an XML sidecar. If that metadata is out of
+        date, the statistics may not correspond to the actual data.
+
+        Additionally, GDAL will save statistics to file metadata as a
+        side effect if that metadata does not already exist.
+
+        """
+        cdef double min, max, mean, std
+        cdef GDALRasterBandH band = NULL
+
+        band = self.band(bidx)
+
+        if clear_cache:
+            IF (CTE_GDAL_MAJOR_VERSION, CTE_GDAL_MINOR_VERSION) >= (3, 2):
+                GDALDatasetClearStatistics(self._hds)
+            ELSE:
+                pass
```

Comment:
If what GDAL does (update or write a file as a side-effect of computing stats) is bad, this (deletion of data in a file) is maybe worse? I'm conflicted about this and would love some feedback.

Reply:
> this (deletion of data in a file) is maybe worse?

I'm not sure to understand ☝️ 

> update or write a file as a side-effect of computing stats

This would only happens if a file is opened with `r+` right? I would be worry if not (specifically for COGs)

Reply:
@vincentsarago it's my understanding that statistics might be saved in a file no matter what the opening mode is. The API docs don't specify, and I assume the behavior is format specific, so we would need to read the implementation.